src/raft/raft.rs

```diff
@@ -627,14 +689,24 @@ impl<T: Storage> Raft<T> {
         self.votes = FlatMap::default();
         let (last_index, max_inflight) = (self.raft_log.last_index(), self.max_inflight);
         let self_id = self.id;
-        for (id, p) in self.mut_prs() {
+
+        let rebuild_progress = |id: u64, p: &mut Progress| {
+            let is_learner = p.is_learner;
             *p = new_progress(last_index + 1, max_inflight);
-            if id == &self_id {
+            p.is_learner = is_learner;
+            if id == self_id {
                 p.matched = last_index;
             }
-        }
+        };
+
         self.pending_conf = false;
         self.read_only = ReadOnly::new(self.read_only.option);
+
+        let prs = self.mut_prs();
+        prs.voters
```

Comment:
I think `ProgressSet` can provide methods `iter_mut`, so you can just write `prs.iter_mut().for_each(...)`.

Reply:
I have tried with `fn iter_mut(&mut self) -> impl Iterator<Item = ...>`. But it can't return a `mut iterator`. So `iter_mut` need to return a `SelfDefinedIterator`, which I don't want to define.

Reply:
Why? Check it out: https://play.rust-lang.org/?gist=cd9c2f64d473b39cece7d5c6060bc6fb&version=nightly